sound/soc/sof/fw-file-profile.c

```diff
@@ -0,0 +1,266 @@
+{
+	bool fw_path_allocated = false;
+	bool fw_lib_path_allocated = false;
+	int ret = 0;
+
+	/* firmware path */
+	if (path_override->fw_path) {
+		out_profile->fw_path = path_override->fw_path;
+	} else if (path_override->fw_path_postfix) {
+		out_profile->fw_path = devm_kasprintf(dev, GFP_KERNEL, "%s/%s",
+							desc->default_fw_path[ipc_type],
+							path_override->fw_path_postfix);
+		if (!out_profile->fw_path)
+			return -ENOMEM;
+
+		fw_path_allocated = true;
+	} else {
+		out_profile->fw_path = desc->default_fw_path[ipc_type];
+	}
+
+	/* firmware library path */
+	if (path_override->fw_lib_path) {
+		out_profile->fw_lib_path = path_override->fw_lib_path;
+	} else if (desc->default_lib_path[ipc_type]) {
+		if (path_override->fw_path_postfix) {
+			out_profile->fw_lib_path = devm_kasprintf(dev,
+							GFP_KERNEL, "%s/%s",
+							desc->default_lib_path[ipc_type],
+							path_override->fw_path_postfix);
+			if (!out_profile->fw_path) {
+				ret = -ENOMEM;
+				goto out;
+			}
+
+			fw_lib_path_allocated = true;
+		} else {
+			out_profile->fw_lib_path = desc->default_lib_path[ipc_type];
+		}
+	}
+
+	/* firmware filename */
+	if (path_override->fw_name)
+		out_profile->fw_name = path_override->fw_name;
+	else
+		out_profile->fw_name = desc->default_fw_filename[ipc_type];
+
+	if (path_override->fw_path_postfix)
+		out_profile->fw_path_postfix = path_override->fw_path_postfix;
+
+	/* topology path */
+	if (path_override->tplg_path)
+		out_profile->tplg_path = path_override->tplg_path;
+	else
+		out_profile->tplg_path = desc->default_tplg_path[ipc_type];
+
+	/* topology name */
+	if (path_override->tplg_name)
+		out_profile->tplg_name = path_override->tplg_name;
+
+	out_profile->ipc_type = ipc_type;
+
+	/* Test only default firmware file */
+	if (!path_override->fw_path && !path_override->fw_name)
+		ret = sof_test_file_profile(dev, out_profile);
+
+out:
+	if (ret) {
+		/* Free up path strings created with devm_kasprintf */
+		if (fw_path_allocated)
+			devm_kfree(dev, out_profile->fw_path);
+		if (fw_lib_path_allocated)
+			devm_kfree(dev, out_profile->fw_lib_path);
+
+		memset(out_profile, 0, sizeof(*out_profile));
+	}
+
+	return ret;
+}
+
+static void
+sof_missing_firmware_notification(struct device *dev,
+				  const enum sof_ipc_type ipc_type,
+				  const struct sof_dev_desc *desc,
+				  struct sof_loadable_file_profile *path_override)
+{
+	char *marker;
+	int i;
+
+	dev_err(dev, "No SOF firmware file was found.\n");
+
+	for (i = 0; i < SOF_IPC_TYPE_COUNT; i++) {
+		if (i == ipc_type)
+			marker = "Requested";
+		else
+			marker = "Fallback";
+
+		if (path_override->fw_path_postfix)
+			dev_err(dev, " %s file: %s/%s/%s\n", marker,
+				desc->default_fw_path[i],
+				path_override->fw_path_postfix,
+				desc->default_fw_filename[i]);
+		else
+			dev_err(dev, " %s file: %s/%s\n", marker,
+				desc->default_fw_path[i],
+				desc->default_fw_filename[i]);
+	}
+
+	dev_err(dev, "Check if you have 'sof-firmware' package installed.\n");
```

Comment:
Minor nit: not sure how stable "sof-firmware" is as a pointer. In many RPM-based distros, package name is alsa-sof-firmware. Thus we just referred to sof-bin in the kernel error in the past.

Reply:
@ujfalusi Yup, we could add a "If you are seeing this message, you are looking too close." disclaimer here. ;)